### PR TITLE
KConfig Set

### PR DESCRIPTION
**Pull request by darddan**
## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy.
- [x] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

### CODE REVIEW

**Comment by darddan** [comment]
For some reason I cannot add a `Work in progress` label. 
Also I have a question. @FelixResch noticed that if you add a key into the GUI backend with brackets (example `[key name]`) it will be saved so into the file. Should I throw an error if I encounter such keys or should it be implemented in other ways.

**Comment by markus2330** [comment]
> Also I have a question. @FelixResch noticed that if you add a key into the GUI backend with brackets (example [key name]) it will be saved so into the file. Should I throw an error if I encounter such keys or should it be implemented in other ways.

I think it is a rather low-priority issue. You can stay compatible to how it was before. If it is not too much work, proper escaping would be the best way. But then you need to somehow distinguish which `[]` are for locales (unescaped) and which `[]``are to be escaped.

**Comment by darddan** [comment]
jenkins build libelektra please

**Comment by darddan** [comment]
Is it possible to somehow get this file `/home/jenkins/workspace/libelektra_PR-3277/build directory/Testing/Temporary/MemoryChecker.*.log`

**Comment by mpranj** [comment]
Yep. In the jenkins blueocean UI click on the artifacts tab. The output is part of [`artifacts/debian-buster-full/DynamicAnalysis.xml`](https://build.libelektra.org/job/libelektra/job/PR-3277/5/artifact/artifacts/debian-buster-full/DynamicAnalysis.xml).

If something fails only on the buildserver and you can't reproduce it locally you can also add more artifacts by editing `/scripts/jenkins/Jenkinsfile` to debug the situation. Basically you can add anything from the build directory. Do this only when needed as it can store tons of data on the buildserver.

**Comment by darddan** [comment]
Thanks I also noticed that the report was being shown in another CI that it also failed

**Comment by markus2330** [comment]
jenkins build libelektra please

File: src/plugins/kconfig/README.md [hunk 1]
```diff
@@ -73,6 +74,13 @@ echo 'key=Value' > `kdb file /tests/kconfig`
 kdb get /tests/kconfig/key
 #> Value
 
+# Set the value to Example
+kdb set /tests/kconfig/key Example
+
+# Verify that the value has changed in the file too
+cat `kdb file /tests/kconfig`
+#> key=Example
+
```
Comment (markus2330) [inline src/plugins/kconfig/README.md]: Also give examples for locales.

File: src/plugins/kconfig/testmod_kconfig.cpp [hunk 2]
```diff
@@ -99,6 +100,57 @@ TEST (kconfig, simple_file_get)
...
+	std::string tmp;
+
+	getline (file, tmp);
+	succeed_if_same (tmp, "[group]", "");
+	getline (file, tmp);
+	succeed_if_same (tmp, "key[de]=Hallo", "");
+	getline (file, tmp);
+	succeed_if_same (tmp, "key[en][$e][$i]=Hello", "");
+	getline (file, tmp);
+	succeed_if_same (tmp, "title=KConfig Test", "");
```
Comment (markus2330) [inline src/plugins/kconfig/testmod_kconfig.cpp]: A few more tests would not harm, e.g. with umlauts, escaping, longer names.

**Comment by darddan** [comment]
jenkins build libelektra please

**Comment by darddan** [comment]
jenkins build libelektra please

**Comment by mpranj** [comment]
@darddan you need to fix the formatting, otherwise the jenkins tests will abort early.

If you have the all the formatting tools in the correct versions installed locally, you can reformat using `ctest -VV -R testscr_check_formatting`. If not, the buildserver also gives a patch including instructions on how to apply the patch.

**Comment by darddan** [comment]
jenkins build libelektra please

**Comment by darddan** [comment]
@markus2330 @mpranj I reran the `FULL` job, it builds correctly, but the unsuccessful one still shows up here (as duplicate). could you tell me what to do next, is there something I'm overlooking?

**Comment by darddan** [comment]
The last one timed out. But looking at the report [here](https://cirrus-ci.com/task/5978845924032512),  this passed 6 times already.

**Comment by mpranj** [comment]
Thank you very much for implementing the write functionality!

Looks like the `FULL` job was stuck (maybe a cirrus-ci bug) and you resolved it by pushing another commit.

**Comment by markus2330** [comment]
I added the kconfig plugin in the Debian package 0c83880ab84a85ac7b983a0094c924a957674b71

@mpranj did you test the plugin with your (old) KDE config? I plan to do it once the plugin is in the Debian package.

**Comment by mpranj** [comment]
I did not have time to do any manual tests. I don't think I have any interesting KDE config, I basically only use KDevelop in gnome.

I'll have to test it with a fresh KDE installation.

**Pull request merged**